alf/algorithms/one_step_loss.py

```diff
@@ -41,3 +44,139 @@ def __init__(self,
+            # [T, B, 1]
+            assert info.reward.shape[-1] == 1, (
+                "Cannot handle multi-dimensional reward for OneStepTDQRLoss.")
+
+        value_shape = target_value.shape
+        rewards = info.reward.unsqueeze(-1).expand(value_shape)
+        step_types = info.step_type.unsqueeze(-1).expand(value_shape)
+        discounts = info.discount.unsqueeze(-1).expand(
+            value_shape) * self._gamma
+
+        returns = value_ops.one_step_discounted_return(
+            rewards=rewards,
+            values=target_value,
+            step_types=step_types,
+            discounts=discounts)
+
+        return returns
+
+    def forward(self, info, value, target_value):
+        """Calculate the loss.
+
+        The first dimension of all the tensors is time dimension and the second
+        dimesion is the batch dimension.
+
+        Args:
+            info (namedtuple): experience collected from ``unroll()`` or
+                a replay buffer. All tensors are time-major. ``info`` should
+                contain the following fields:
+                - reward:
+                - step_type:
+                - discount:
+            value (torch.Tensor): the time-major tensor for the value at each time
+                step. The loss is between this and the calculated return.
+            target_value (torch.Tensor): the time-major tensor for the value at
+                each time step. This is used to calculate return. ``target_value``
+                can be same as ``value``.
+        Returns:
+            LossInfo: with the ``extra`` field same as ``loss``.
+        """
+        assert value.ndim == 3 and target_value.ndim == 3, (
+            "input value and target_value should have shape (T, B, N)")
+        returns = self.compute_td_target(info, target_value)
+        value = value[:-1]
+
+        cdf_midpoints = self._cdf_midpoints.view(1, 1, 1, -1)
+        quantiles = value.unsqueeze(-2)
+        quantiles_target = returns.detach().unsqueeze(-1)
+        diff = quantiles_target - quantiles
+
+        if self._debug_summaries and alf.summary.should_record_summaries():
+            mask = info.step_type[:-1] != StepType.LAST
+            with alf.summary.scope(self._name):
+
+                def _summarize(v, r, qtd, suffix):
+                    alf.summary.scalar(
+                        "explained_variance_of_return_by_value" + suffix,
+                        tensor_utils.explained_variance(v, r, mask))
+                    safe_mean_hist_summary('values' + suffix, v, mask)
+                    safe_mean_hist_summary('returns' + suffix, r, mask)
+                    safe_mean_hist_summary("quantile_td" + suffix, qtd, mask)
+
+                for i in range(value.shape[-1]):
+                    for j in range(returns.shape[-1]):
+                        suffix = f'/{j}-{i}'
+                        _summarize(value[..., i], returns[..., j],
+                                   diff[..., j, i], suffix)
```

Comment:
For num_quantiles, this will generate 4 * 2500 summaries. Too much!
Also, value[..., i] and returns[..., j] are summarized multiple times.

Since values are quantile estimations, it would be nice if we can summarize the percent of target that is smaller than the value so that we can get an idea whether the quantiles are estimated correctly.

Reply:
Now only summarize a mean explained cdf of target by value quantile as you suggested.